### DISCUSSION TITLE
(Done) March 12th Meetup: Decompilation and clang plugins

### DISCUSSION THREAD

**Discussion by frenchy64** [madclj/madclj.com#12] [Announcements]
👍 to RSVP

https://madclj.com/

# Location

Wednesday, March 12th, 2025
6:30 PM to 9:00 PM CT

[StartingBlock Madison](https://www.google.com/maps/search/?api=1&query=43.081245%2C%20-89.374054)
821 E Washington Ave 2nd floor · Madison, WI

# Schedule

Sayan will give a presentation on a clang plugin she is designing.

Ambrose will demonstrate how to to decompile Clojure to inspect the equivalent Java code. We will use this to show how `let{fn}` is implemented in Clojure, then compare this to jank's actual LLVM IR output.

6:30pm arrive, food & drink provided
7pm start meetup
9pm end meetup

**Comment by brookeswanson** [comment]
I shall be there!

**Comment by frenchy64** [comment]
Any ideas for topics? Would people be up for taking turns to plug into the projector and show a thing or two they've been working on?

**Reply by sivakusayan** [reply]
Maybe we could do a papers we love type thing?    

If that doesn't work, I could do a rehash of a presentation I gave at Singlewire at the end of January, although James Elliot would have seen it already 🙂 it was a ~50 minute presentation that fleshes out some mathematical ideas behind the clang plugin I want to write, although ~15 minutes of that was computer architecture background people here are probably more familiar with.

Otherwise, I've been trying to figure out how lock-free programming works recently, and assuming I can learn the (very basics) of it in a month and try to do some implementation, I could prepare a presentation for what I learned so far in May. Although, I'm moving to Milwaukee soon, so not sure if I'll still be here in May.

**Comment by ericnormand** [comment]
I'll be there with the girls!

**Comment by sivakusayan** [comment]
Did we decide on a schedule for tomorrow? Or were we thinking of skipping this month?

**Reply by frenchy64** [reply]
Still on, but haven't decided what to do exactly yet.

**Reply by frenchy64** [reply]
@sivakusayan could you do a 30 minute version of your clang talk?

**Reply by sivakusayan** [reply]
Sounds good :)

**Comment by frenchy64** [comment]
@madclj/meetup tomorrow's meetup will feature two talks:

Sayan will give a presentation on a clang plugin she is designing.

Ambrose will demonstrate how to to decompile Clojure to inspect the equivalent Java code. We will use this to show how `let{fn}` is implemented in Clojure, then compare this to jank's actual LLVM IR output.

**Comment by frenchy64** [comment]
Pizza is ordered. I should be in the lobby at 6:30pm to let people in.

**Reply by sivakusayan** [reply]
I'm downstairs!

**Comment by frenchy64** [comment]
Some relevant thoughts to Sayan's talk wrt jank:
- jank objects are structs whose first members are a `uint8_t` tag. there is no subclassing, instead objects are cast from the [`object`](https://github.com/jank-lang/jank/blob/f04683c7f3f3bb1b03b786d7906f2d3462a80ea7/compiler%2Bruntime/include/cpp/jank/runtime/object.hpp#L219-L222) struct---which just has a tag---to other structs that also start with a tag with a big [`switch` statement](https://github.com/jank-lang/jank/blob/f04683c7f3f3bb1b03b786d7906f2d3462a80ea7/compiler%2Bruntime/include/cpp/jank/runtime/visit.hpp#L78) to simulate polymorphism.
- [this](https://github.com/jank-lang/jank/blob/f04683c7f3f3bb1b03b786d7906f2d3462a80ea7/compiler%2Bruntime/src/cpp/jank/codegen/llvm_processor.cpp#L214) llvm optimization pass can automatically convert `struct { T field}` into `T` (apparently that's subtly different). It's used to optimize closures that close over 1 variable like `(let [v 1] (fn [] v))`. The extra argument passed to initialize the closure is just `1` instead of `struct { v=1 }`.
